src/AzureAuth/Ado/AdoToken.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AzureAuth.Ado
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    using Microsoft.Office.Lasso.Interfaces;
+
+    /// <summary>
+    /// Business logic for either getting an ADO PAT from <see cref="IEnv"/>
+    /// or doing normal AAD Authentication.
+    /// </summary>
+    public static class AdoToken
+    {
+        private static readonly IEnumerable<string> PatEnvVars = new[]
+        {
+            EnvVars.AdoPat,
+            EnvVars.SystemAccessToken,
+        };
+
+        /// <summary>
+        /// Get either an ADO PAT from the environment. If one is not set, then do normal ADO AAD Auth.
+        /// </summary>
+        /// <param name="env">The <see cref="IEnv"/> to use for interacting with the environment.</param>
+        /// <returns>A PAT from the env if set, otherwise null.</returns>
+        public static PatResult PatFromEnv(IEnv env)
+        {
+            foreach (var envVar in PatEnvVars)
+            {
+                var pat = env.Get(envVar);
+                var exists = !string.IsNullOrEmpty(pat);
+
+                if (exists)
+                {
+                    return new ()
+                    {
+                        Exists = exists,
+                        Value = pat,
+                        EnvVarSource = envVar,
+                    };
+                }
+            }
+
+            return new ()
+            {
+                Exists = false,
+            };
+        }
+
+        /// <summary>
+        /// The result struct for retrieving a PAT out of the environment.
+        /// </summary>
+        public record PatResult
+        {
+            /// <summary>
+            /// Gets a value indicating whether a PAT was found in the env.
+            /// </summary>
+            public bool Exists { get; init; }
+
+            /// <summary>
+            /// Gets the source env var that contained the PAT.
+            /// </summary>
+            public string EnvVarSource { get; init; }

```

Comment:
I think `EnvVar` is sufficient and you might not need the `Source` suffix. It's not doing any hard, but I don't know if it's necessary. The `<summary>` should help clarify in the event of confusion.

Reply:
Mmmm, I'm going to Won't Fix this one. I considerd just Source. But I prefer the self-documenting code to having the rely on XML docs because it's not clear enough on it's own.